### ISSUE TITLE
[FEAT] COT Reasoning 을 통한 최종 정답 도출

### ISSUE THREAD

**Issue by jagaldol** [boostcampaitech7/level2-nlp-generationfornlp-nlp-06-lv3#46]
### ⚠️ Please check for similar feature requests

- [X] 이전에 제안된 [기능 요청](https://github.com/boostcampaitech7/level2-nlp-generationfornlp-nlp-06-lv3/issues?q=+is%3Aissue+label%3A%22Type%3A+Enhancement%22+) 중 중복되는 것이 없습니다.

### 💡 기능 요약

Chain Of Thought 를 통해 정답의 이유를 단계별로 생각하도록 설계

### 📄 기능 설명

단순히 답만을 반환하는 방식에서 벗어나 **추론 과정(COT: Chain of Thought)**을 포함한 답변을 생성하고, 이를 JSON 형태로 출력하도록 구현할 계획입니다. JSON 형태의 답변은 파싱하기 쉬우며, 이 방식을 통해 유연하고 일관된 데이터 처리가 가능해집니다.

또한, 추론 과정과 LangChain의 적용은 향후 RAG #40 이슈와도 연관이 있으므로, 최대한 통일성을 유지하며 작업해야 합니다.

#### 1. COT 프롬프트 설계
- [x] **Few-shot COT 프롬프트 고안**: COT 방식의 추론 예제를 포함한 프롬프트 설계
- [x] **JSON 형태로 출력하도록 프롬프트 고안**: COT 및 최종 답변이 명확히 JSON으로 출력되도록 설정

#### 2. 추론 및 LangChain 적용
- [x] **`predict` 로직 개선**: 
  - 기존의 로직을 개선하여 COT 추론과 JSON 출력 방식으로 answer를 추출
- [x] **LangChain 활용**: 
  - Huggingface 모델 객체를 직접 사용하는 방식을 LangChain 기반으로 전환
- [x] **JSONOutputParser 적용**: LangChain의 JSONOutputParser를 사용해 일관된 JSON 형태의 답변 처리 구현

#### 3. 학습된 모델의 테스트 (`train.py`)
- [x] **테스트**: 학습된 모델이 새로운 추론 방식(COT 및 JSON 출력)에서 제대로 동작하는지 확인
- [x] **과적합 문제 해결**: 
  - 현재 학습된 모델이 단순히 숫자 하나만 대답하도록 하였기에 과적합으로 동작하지 않을 가능성이 높음
- [x] **필요 시 학습 데이터 수집**: COT 방식에 적합한 Reasoning이 포함된 데이터셋 준비
- [x] **새로운 학습 방식 고안**: COT 및 JSON 형태의 답변 생성을 학습시키기 위한 새로운 방법 설계

### 🎯 기능 도입의 이점

최근의 많은 논문을 보면, 항상 COT를 적용하였을 때 성능이 유의미하게 오르는 현상을 보여줍니다. 우리의 모델 또한, COT를 적용하여 성능을 올려보고자 합니다. 또한, 모델이 어떠한 사고 과정으로 해당 정답을 선택하였는지 알 수 있어, 모델 특성 분석에도 도움이 될 것이라 사료됩니다.

### ⚠️ 잠재적 문제

현재 단순히 선택지 1~5중 가장 logits값이 높은 숫자의 선택지를 최종 선택하도록 되어 있습니다. 이부분을 모델이 reasoning 및 answer를 대답하도록 바꾸어야하니 추론부분을 뜯어 고칠 필요가 존재합니다.
